File: frontend/src/store/trades.ts
import { Module } from "vuex";
import axios from 'axios';

const tradesModule: Module<any, any> = { 
    state: {
        assets: [],
        symbols: [],
        symbolsChoices: [],
        timeframeChoices: [],
        timeframe: '',
        candleInterval: 0,
        walletAmount: 0,
        belowAverage: 0,
        profitMargin: 0,
        profitAs: '',
        spot: true,
        sandbox: true
    },

    getters: {
        assets: state => state.assets,
        symbols: state => state.symbols,
        symbolsChoices: state => state.symbolsChoices,
        timeframeChoices: state => state.timeframeChoices,
        timeframe: state => state.timeframe,
        candleInterval: state => state.candleInterval,
        walletAmount: state => state.walletAmount,
        belowAverage: state => state.belowAverage,
        profitMargin: state => state.profitMargin,
        profitAs: state => state.profitAs,
        spot: state => state.spot,
        sandbox: state => state.sandbox,
    },

    mutations: {
        SET_ASSETS(state, value) { state.assets = value; },
        SET_SYMBOLS(state, value) { state.symbols = value; },
        SET_SYMBOLS_CHOICES(state, value) { state.symbolsChoices = value; },
        SET_TIMEFRAME_CHOICES(state, value) { state.timeframeChoices = value; },
        SET_TIMEFRAME(state, value) { state.timeframe = value; },
        SET_CANDLE_INTERVAL(state, value) { state.candleInterval = value; },
        SET_WALLET_AMOUNT(state, value) { state.walletAmount = value; },
        SET_BELOW_AVERAGE(state, value) { state.belowAverage = value; },
        SET_PROFIT_MARGIN(state, value) { state.profitMargin = value; },
        SET_PROFIT_AS(state, value) { state.profitAs = value; },
        SET_SPOT(state, value) { state.spot = value; },
        SET_SANDBOX(state, value) { state.sandbox = value; },
    },

    actions: {
        loadTrades(state) {
            axios.get(`/api/v1/trades/`, {headers: {'token': localStorage['sk']}}).then(response => {
                state.commit('SET_SANDBOX', response.data.sandbox);
                state.commit('SET_ASSETS', response.data.assets);
                state.commit('SET_SYMBOLS', response.data.symbols);
                state.commit('SET_SYMBOLS_CHOICES', response.data["symbols-choices"]);                
                state.commit('SET_TIMEFRAME_CHOICES', response.data["timeframe-choices"]);
                state.commit('SET_TIMEFRAME', response.data["timeframe"]);
                state.commit('SET_CANDLE_INTERVAL', response.data["candle-interval"]);
                state.commit('SET_WALLET_AMOUNT', response.data["wallet-amount"]);
                state.commit('SET_BELOW_AVERAGE', response.data["below-average"]);
                state.commit('SET_PROFIT_MARGIN', response.data["profit-margin"]);
                state.commit('SET_PROFIT_AS', response.data["profit-as"]);
                state.commit('SET_SPOT', response.data["spot"]);
                state.commit('SET_EXPECTED_LEVERAGE', response.data["expected-leverage"]);
                state.commit('SET_ACTIVATE_PRICE', response.data["activation-price"]);
                state.commit('SET_IN_GREEN', response.data["in-green"]);
                state.commit('SET_USE_AVERAGE', response.data["use-average"]);
                state.commit('SET_VOLUME_TIME_FRAME', response.data["volume-timeframe"]);
                state.commit('SET_TOTAL_VOLUME', response.data["total-volume"]);
                state.commit('SET_MARGIN_TYPE', response.data["margin-type"]);
                console.log(response.data);
            });
        },
        saveTradeConfig(state) {
            axios.post(`/api/v1/trades/`, {
                sandbox: state.getters.sandbox,
                symbols: state.getters.symbols,
                timeframe: state.getters.timeframe,
                candle_interval: state.getters.candleInterval,
                wallet_amount: state.getters.walletAmount,
                below_average: state.getters.belowAverage,
                profit_margin: state.getters.profitMargin,
                profit_as: state.getters.profitAs,
                spot: state.getters.spot,
                activation_price: state.getters.activatePrice,
                in_green: state.getters.inGreen,
                use_average: state.getters.useAverage,
                expected_leverage: state.getters.expectedLeverage,
                volume_timeframe: state.getters.volumeTimeFrame,
                total_volume: state.getters.totalVolume,
                margin_type: state.getters.marginType,
            },{headers: {'token': state.getters.token}});
        }
    }
}

export default tradesModule;

File: backend/views/api/trades.py
from flask import render_template, jsonify, current_app, request
from flask_classful import FlaskView, route

from backend.utils.auth import login_required
from backend import db, pynance

class TradesApiView(FlaskView):
    
    decorators = [ login_required ]

    def get(self):
        """Loads the current configuration and loads it into the store in the frontend.
        The user can change the values and save the changes to confirm the changes.

        Returns:
            [dict]: {
                'assets': [A list of available assets from Binance],
                'symbols': [The selected symbols to trade],
                'symbols-choices': [A list of available symbols from Binance],
                'timeframe-choices': [A list of available graph time frames],
                'timeframe': Returns a string representing the selected timeframe,
                'candle-interval': Returns an integer which represents the selected candle-interval,
                'wallet-amount': Returns an integer which is the total amount of % available free coins to use when placing buy orders,
                'below-average': Returns an integer which pushes the lowest average x% lower,
                'profit-margin': Returns an integer which represent the profit % used for sell orders,
                'profit-as': Is used to calculate all profit as the selected asset
            }
        """
        symbols = [i['symbol'] for i in pynance.assets.symbols().json]
        assets = [i['coin'] for i in pynance.wallet.balance().json]
        from backend.models.config import ConfigModel
        config = ConfigModel.query.first()
        return jsonify({
            'sandbox': config.sandbox,
            'assets': assets,
            'symbols': config.symbols,
            'symbols-choices': symbols,
            'timeframe-choices': ['1m', '3m', '5m', '15m', '30m', '1h', '2h', '4h', '6h', '8h', '12h', '1d', '3d', '1w', '1M'],
            'timeframe': config.timeframe,
            'candle-interval': config.candle_interval,
            'wallet-amount': config.wallet_amount,
            'below-average': config.below_average,
            'profit-margin': config.profit_margin,
            'profit-as': config.profit_as,
            'spot': config.spot,
            'default-stop-loss': config.default_stop_loss,
            'total-TP': config.total_tp,            
            'in-green': config.in_green,
            'move-stop-loss': config.move_stop_loss,
            'take-profit': config.take_profit,
            'use-average': config.use_average,
            'expected-leverage': config.expected_leverage,
            'volume-timeframe': config.volume_timeframe,
            'total-volume': config.total_volume,
            'margin-type': config.margin_type,
        }), 200

    def post(self):
        """Saves the users trade configuration which is used while trading
        """
        from backend.models.bot import BotModel
        bot = BotModel.query.first()
        bot.update_data({'online': False})
        data = request.json
        from backend.models.config import ConfigModel
        config = ConfigModel.query.first()
        config.update_data({**data})
        [i.set_active(False) for i in bot.orders if i.symbol not in config.symbols]
        [i.set_active(True) for i in bot.orders if i.symbol in config.symbols and i.sold_for == 0]
        if not config.sandbox: config.remove_sandbox()
        return jsonify({}), 200
